### ISSUE TITLE
Inferred type can be too generic with `number` and `Int`

### ISSUE THREAD

**Issue by rtavenner** [elm-lang/elm-compiler#1605]
In this code, the compiler infers that the ```toInt``` function has type ```number -> Int```, when it should have type ```Int -> Int```. As a result, ```onePointFive``` has type ```Int``` and value ```1.5```! This behavior can be seen in the Try Elm editor, in the Ellie-app editor, and in my installed copy of Elm (version 18.0).
```
import Html

intId : Int -> Int
intId x = x

toInt x =
    if x > 0
    then intId x
    else intId x

onePointFive : Int
onePointFive = toInt 1.5

main = Html.text (toString onePointFive)
```
Interestingly, if I add the inferred type signature (```toInt : number -> Int```) to the ```toInt``` function, the code does not compile, but not with the expected error. Instead, it reports:
```
The left argument of (>) is causing a type mismatch.
(>) is expecting the left argument to be a:
    comparable
But the left argument is:
    number
Hint: Only ints, floats, chars, strings, lists, and tuples are comparable.
```
However, if I instead add the signature ```toInt : comparable -> Int```, it compiles successfully! 
Then (and only then, this won't compile before you add that type signature), you can do crazy things like 
```
banana : Int
banana = toInt "banana"
```
and wreak all sorts of havoc.

So, to recap: 

- The wrong type signature was inferred.
- Copying the inferred type signature into the program caused a compile error, but not the one you would expect.
- The error it gave me led me to try an even less sensible type signature, which somehow compiled perfectly, despite being neither the correct signature or the inferred signature. 
- This new type signature, once added, allowed even weirder things to compile than before it was added.
- This allowed the construction of the nonsensical value ```"banana" : Int```.

I don't know if it matters, but my browser is Firefox 54.0, and my computer is running Ubuntu 16.10.

**Comment by evancz** [comment]
In the latest build I am getting:

```
-- TYPE MISMATCH ------------------------------------------------------ temp.elm

The 1st argument to `toInt` is not what I expect:

12| onePointFive = toInt 1.5
                         ^^^
This argument is a float of type:

    Float

But `toInt` needs the 1st argument to be:

    Int

Note: Read <https://elm-lang.org/0.19.0/implicit-casts> to learn why Elm does
not implicitly convert Ints to Floats. Use toFloat and round to do explicit
conversions.
```

Also an error with the `"banana"` example. Thanks for the report!
